### ISSUE TITLE
[Regression] Commit e70dda2 broke a Pomelo.MySql test

### ISSUE THREAD

**Issue by lauxjpn** [mysql-net/MySqlConnector#713]
The following test in the `GearsOfWarQueryTestBase` class worked before the e70dda2 commit:

```c#
[ConditionalFact]
public virtual void Select_null_conditional_with_inheritance_negative()
{
    using (var context = CreateContext())
    {
        var query = context.Factions
            .Where(f => f is LocustHorde)
            .Select(f => EF.Property<string>((LocustHorde)f, "CommanderName") != null
                ? ((LocustHorde)f).Eradicated
                : null);

        var result = query.ToList();
        Assert.Equal(2, result.Count);
        Assert.True(result.Contains(true));
        Assert.True(result.Contains(false));
    }
}
```
It generates the following SQL:
```sql
SELECT CASE
    WHEN `f`.`CommanderName` IS NOT NULL THEN `f`.`Eradicated`
    ELSE NULL
END
FROM `Factions` AS `f`
WHERE (`f`.`Discriminator` = 'LocustHorde') AND (`f`.`Discriminator` = 'LocustHorde')
```
With the `Fractions` table defined like this:
```sql
CREATE TABLE `Factions` (
  `Id` int(11) NOT NULL,
  `Name` longtext,
  `CapitalName` varchar(255) DEFAULT NULL,
  `Discriminator` longtext NOT NULL,
  `CommanderName` varchar(255) DEFAULT NULL,
  `Eradicated` bit(1) DEFAULT NULL,
  PRIMARY KEY (`Id`),
  UNIQUE KEY `IX_Factions_CommanderName` (`CommanderName`),
  KEY `IX_Factions_CapitalName` (`CapitalName`),
  CONSTRAINT `FK_Factions_Cities_CapitalName` FOREIGN KEY (`CapitalName`) REFERENCES `cities` (`Name`) ON DELETE RESTRICT,
  CONSTRAINT `FK_Factions_LocustLeaders_CommanderName` FOREIGN KEY (`CommanderName`) REFERENCES `locustleaders` (`Name`) ON DELETE RESTRICT
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_0900_ai_ci;
```
Now it throws the following exception:
```
[xUnit.net 00:05:05.16]     Pomelo.EntityFrameworkCore.MySql.FunctionalTests.Query.GearsOfWarQueryMySqlTest.Select_null_conditional_with_inheritance_negative [FAIL]
  X Pomelo.EntityFrameworkCore.MySql.FunctionalTests.Query.GearsOfWarQueryMySqlTest.Select_null_conditional_with_inheritance_negative [9ms]
  Error Message:
   System.FormatException : One of the identified items was in an invalid format.
  Stack Trace:
     at MySqlConnector.Core.Row.ParseUInt64(ReadOnlySpan`1 data) in C:\projects\mysqlconnector\src\MySqlConnector\Core\Row.cs:line 460
   at MySqlConnector.Core.Row.ReadBit(ReadOnlySpan`1 data, ColumnFlags columnFlags) in C:\projects\mysqlconnector\src\MySqlConnector\Core\Row.cs:line 450
   at MySqlConnector.Core.TextRow.GetValueCore(ReadOnlySpan`1 data, ColumnDefinitionPayload columnDefinition) in C:\projects\mysqlconnector\src\MySqlConnector\Core\TextRow.cs:line 109
   at MySqlConnector.Core.Row.GetValue(Int32 ordinal) in C:\projects\mysqlconnector\src\MySqlConnector\Core\Row.cs:line 32
   at MySqlConnector.Core.Row.GetBoolean(Int32 ordinal) in C:\projects\mysqlconnector\src\MySqlConnector\Core\Row.cs:line 67
   at lambda_method(Closure , QueryContext , DbDataReader , ResultContext , Int32[] , ResultCoordinator )
   at Microsoft.EntityFrameworkCore.Query.RelationalShapedQueryCompilingExpressionVisitor.QueryingEnumerable`1.Enumerator.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.GearsOfWarQueryTestBase`1.Select_null_conditional_with_inheritance_negative()
```
Which happens at the following line, that was added in e70dda2:
https://github.com/mysql-net/MySqlConnector/blob/9b8f3a4d8259144b52990d804f1dca303314289f/src/MySqlConnector/Core/Row.cs#L453-L454

So this column is likely the issue:
```sql
`Eradicated` bit(1) DEFAULT NULL
```
Of 13681 executed tests, this is the only one suddenly failing.

See https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/849 for the full CI output.

**Comment by lauxjpn** [comment]
For this query, `columnFlags = Unsigned | Binary` with `data` containing only 1 `byte` (with the value `1`).
So it is assumed a `ulong`, but the data contains only a `byte`.
https://github.com/mysql-net/MySqlConnector/blob/9b8f3a4d8259144b52990d804f1dca303314289f/src/MySqlConnector/Core/Row.cs#L436-L454

This is how it looks in the debugger:

![issue713-01](https://user-images.githubusercontent.com/14178357/66203618-16545d80-e6a9-11e9-83f6-0cd99b9b3738.png)

**Comment by bgrainger** [comment]
When I execute this query in MySQL Workbench (with a single row in the table where `CommanderName` is `'a'` and `Eradicated` is `1`), it displays `0` as the result of the query (not `1`).

So this looks like it might be a bug in MySQL Server.

Even if MySqlConnector caught the FormatException, it would probably fall back to returning `0`, just like MySQL Workbench. Wouldn't that still break the Pomelo integration test, which is expecting `1`?

**Comment by lauxjpn** [comment]
It would. Let me recheck the test behavior before `0.59.1`.

**Comment by bgrainger** [comment]
The correct result is returned (in MySQL Workbench and MySqlConnector, where it returns `1m`) with MariaDB 10.4.

This confirms my suspicion that it's a MySQL Server 8.0.17 issue.

**Comment by bgrainger** [comment]
Pre 0.59.1 would have always read the `1` as binary, which would have succeeded. 0.59.1 added the ability to read as binary or text (which prevented `'1'` being read as 49).

**Comment by lauxjpn** [comment]
It fails too for 5.7.27 but succeeds for 5.6.45.

**Comment by lauxjpn** [comment]
Is there a workaround that can be implemented in MySqlConnector until the regression is fixed in MySQL?

**Comment by bgrainger** [comment]
We could look for byte strings that contain bytes other than ASCII digits (0x30 - 0x39) and interpret them as binary. 

This heuristic would fail for numbers that happen to be made up of bytes that are ASCII digits, e.g., `825373492` = `0x31323334` = `'1234'` (or, perhaps more realistically, `49` = `0x31` = `'1'`). This is probably unlikely to happen in practice, but would be a very difficult bug to track down (the value 49 might sometimes be read as `49`, but sometimes as `1` depending on whether subqueries were involved, etc.).

**Comment by bgrainger** [comment]
A safer heuristic, for `BIT(1)` columns:

* `0x00` and `0x30` represent `0`
* `0x01` and `0x31` represent `1`

`0x30` and `0x31` are out-of-range for a `BIT(1)` column, so it's safe to assume they're caused by a server bug.

**Comment by lauxjpn** [comment]
Looks good to me!

Also, does `data` only contain the data for the current column? If it does, if it has less then `64 bits / 8` bytes, it cannot be a `UInt64` and can be interpreted as a MSB byte array.

**Comment by bgrainger** [comment]
Another wrinkle: MySQL 5.6 returns the `CASE WHEN ... END` column as `BIT(2)` . We can extend the heuristic above to handle this case.

**Comment by lauxjpn** [comment]
Interesting, is that because it is nullable and needs another bit to signal the `NULL` case?

**Comment by bgrainger** [comment]
No... it just seems to add another BIT. (`NULL` is indicated by another mechanism.) A `BIT(2)` column is sent as `BIT(3)` on the wire...

We can safely handle up to `BIT(5)` because `0x1F` (maximum value in binary) has no overlap with `'1'` - `'9'`.

**Comment by lauxjpn** [comment]
Well what is better than a bit of hacking on a Friday.

Do you want me to report this to the MySQL bug tracker and add a reference to this issue here, or do you want to do it yourself and provide your technical insight to the matter?

**Comment by bgrainger** [comment]
I was planning to report it (later).

**Comment by bgrainger** [comment]
Fixed in [0.59.2](https://github.com/mysql-net/MySqlConnector/releases/tag/0.59.2).

**Comment by bgrainger** [comment]
Reported to Oracle: https://bugs.mysql.com/bug.php?id=97113
